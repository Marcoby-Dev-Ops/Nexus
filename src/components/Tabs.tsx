import React, { useState } from 'react';
import PropTypes from 'prop-types';

/**
 * @interface Tab
 * @description Tab item for the Tabs component.
 */
export interface Tab {
  label: string;
  content: React.ReactNode;
}

/**
 * @interface TabsProps
 * @description Props for the Tabs component.
 */
export interface TabsProps {
  tabs: Tab[];
  className?: string;
}

/**
 * @name Tabs
 * @description A tabs component for organizing content.
 * @param {TabsProps} props - The props for the component.
 * @returns {JSX.Element} The rendered Tabs component.
 */
export const Tabs: React.FC<TabsProps> = React.memo(({ tabs, className = '' }) => {
  const [active, setActive] = useState(0);
  return (
    <div className={className}>
      <div className="flex border-b border-border mb-2">
        {tabs.map((tab, idx) => (
          <button
            key={tab.label}
            className={`px-4 py-2 -mb-px border-b-2 font-medium transition-colors ${active === idx ? 'border-primary text-primary' : 'border-transparent text-muted-foreground hover:text-foreground'}`}
            onClick={() => setActive(idx)}
            aria-selected={active === idx}
            role="tab"
            tabIndex={active === idx ? 0 : -1}
          >
            {tab.label}
          </button>
        ))}
      </div>
      <div role="tabpanel">{tabs[active].content}</div>
    </div>
  );
});

Tabs.displayName = 'Tabs';

Tabs.propTypes = {
  tabs: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      content: PropTypes.node.isRequired,
    })
  ).isRequired,
  className: PropTypes.string,
};

export default Tabs; 