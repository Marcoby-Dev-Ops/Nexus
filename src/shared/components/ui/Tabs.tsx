import React, { createContext, useContext, useState } from 'react';
import { cn } from '@/shared/utils/styles';

interface TabsContextValue {
  value: string;
  onValueChange: (value: string) => void;
}

const TabsContext = createContext<TabsContextValue | undefined>(undefined);

export interface TabsProps {
  defaultValue?: string;
  value?: string;
  onValueChange?: (value: string) => void;
  children: React.ReactNode;
  className?: string;
}

export interface TabsListProps extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode;
}

export interface TabsTriggerProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  value: string;
  children: React.ReactNode;
}

export interface TabsContentProps extends React.HTMLAttributes<HTMLDivElement> {
  value: string;
  children: React.ReactNode;
}

/**
 * @name Tabs
 * @description A tabs container component
 */
export const Tabs: React.FC<TabsProps> = ({
  defaultValue,
  value: controlledValue,
  onValueChange,
  children,
  className,
}) => {
  const [internalValue, setInternalValue] = useState(defaultValue || '');
  const value = controlledValue !== undefined ? controlledValue: internalValue;
  
  const handleValueChange = (newValue: string) => {
    if (controlledValue === undefined) {
      setInternalValue(newValue);
    }
    onValueChange?.(newValue);
  };

  return (
    <TabsContext.Provider value={{ value, onValueChange: handleValueChange }}>
      <div className={cn("w-full", className)}>
        {children}
      </div>
    </TabsContext.Provider>
  );
};

/**
 * @name TabsList
 * @description A container for tab triggers
 */
export const TabsList: React.FC<TabsListProps> = ({ children, className, ...props }) => {
  return (
    <div
      className={cn(
        "inline-flex items-center justify-start space-x-2 rounded-md bg-muted px-2 py-1 text-muted-foreground",
        className
      )}
      {...props}
    >
      {children}
    </div>
  );
};

/**
 * @name TabsTrigger
 * @description A tab trigger button
 */
export const TabsTrigger: React.FC<TabsTriggerProps> = ({ 
  value, 
  children, 
  className, 
  ...props 
}) => {
  const context = useContext(TabsContext);
  if (!context) {
    throw new Error('TabsTrigger must be used within Tabs');
  }

  const isActive = context.value === value;

  return (
    <button
      type="button"
      className={cn(
        "inline-flex items-center justify-center whitespace-nowrap rounded-sm px-4 py-2 text-sm font-medium ring-offset-background transition-all focus-visible: outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50",
        isActive
          ? "bg-background text-foreground shadow-sm"
          : "hover: bg-muted hover:text-foreground",
        className
      )}
      onClick={() => context.onValueChange(value)}
      {...props}
    >
      {children}
    </button>
  );
};

/**
 * @name TabsContent
 * @description Tab content container
 */
export const TabsContent: React.FC<TabsContentProps> = ({ 
  value, 
  children, 
  className, 
  ...props 
}) => {
  const context = useContext(TabsContext);
  if (!context) {
    throw new Error('TabsContent must be used within Tabs');
  }

  if (context.value !== value) {
    return null;
  }

  return (
    <div
      className={cn(
        "mt-2 ring-offset-background focus-visible: outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2",
        className
      )}
      {...props}
    >
      {children}
    </div>
  );
};

export default Tabs; 