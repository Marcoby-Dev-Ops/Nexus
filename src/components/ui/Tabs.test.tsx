/**
 * @file Tabs.test.tsx
 * @description Unit and snapshot tests for the Tabs component.
 */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Tabs from './Tabs';

describe('Tabs', () => {
  const tabs = [
    { label: 'Tab 1', content: <div>Content 1</div> },
    { label: 'Tab 2', content: <div>Content 2</div> },
  ];

  it('renders tab labels', () => {
    render(<Tabs tabs={tabs} />);
    expect(screen.getByText('Tab 1')).toBeInTheDocument();
    expect(screen.getByText('Tab 2')).toBeInTheDocument();
  });

  it('shows content for selected tab', () => {
    render(<Tabs tabs={tabs} />);
    fireEvent.click(screen.getByText('Tab 2'));
    expect(screen.getByText('Content 2')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Tabs tabs={tabs} />);
    expect(asFragment()).toMatchSnapshot();
  });
}); 