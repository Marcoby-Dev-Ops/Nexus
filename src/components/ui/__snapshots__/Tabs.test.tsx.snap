// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs matches snapshot 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="flex border-b border-border mb-2"
    >
      <button
        aria-selected="true"
        class="px-4 py-2 -mb-px border-b-2 font-medium transition-colors border-primary text-primary"
        role="tab"
        tabindex="0"
      >
        Tab 1
      </button>
      <button
        aria-selected="false"
        class="px-4 py-2 -mb-px border-b-2 font-medium transition-colors border-transparent text-muted-foreground hover:text-foreground"
        role="tab"
        tabindex="-1"
      >
        Tab 2
      </button>
    </div>
    <div
      role="tabpanel"
    >
      <div>
        Content 1
      </div>
    </div>
  </div>
</DocumentFragment>
`;
